rfcs/2020-08-31-3645-graphql-api.md

```diff
@@ -0,0 +1,297 @@
+
+* A known schema for errors. Snafu compatibility for `FieldResult<T>` custom errors.
+
+### Later -- Authorization
+
+Initial Vector observability will be single-instance, and available to anyone that has access to the configured port.
+
+Locking down access will initially rely on network configuration.
+
+Later, as we move into multi-instance observability and more granular API permissions, the requirement for user auth and persistence will surface.
+
+While those concerns are out-of-scope for this RFC, choosing a protocol that facilitates authentication is important to avoid backing ourselves into a corner.
+
+GraphQL is not opinionated with auth. We have any authorization mechanism available to us at the intersection of HTTP and WebSockets.
+
+In the previous Timber project, we appended an `Authorization: Bearer <jwt>` header for queries/mutations.
+
+For subscription, we passed a JWT along with the initial WebSocket connection payload; browsers pass limited headers with `Upgrade` requests, so this provided a neat approach to sidestep the lack of a comparable header for WebSockets. The JWT persisted for the life of the open WS.
+
+I anticipate doing something similar with the Vector API. async-graphql has a [Context](https://async-graphql.github.io/async-graphql/en/context.html) struct, which is typically used for passing in shared resources such as a database connection pool, or request-specific data such as the current user session.
+
+This is largely TBD, but the basic mechanisms are there to allow for flexible auth when we need it.
+
+## Doc-level Proposal
+
+A Vector observability layer has already been agreed internally. Work is underway.
+
+This proposal discusses the protocol which will govern communications between a running Vector instance, and a web UI and CLI.
+
+The result of this proposal won't directly impact user interaction with observability tooling.
+
+## Rationale
+
+We have built a decent body of experience with GraphQL at Timber, albeit on non-Vector projects.
+
+The tooling I am proposing here represents the same stack (save for swapping Apollo for Urql.)
+
+In a previous Timber project, we used [gqlgen](https://github.com/99designs/gqlgen) as our server library, written in Go. The tooling with async-graphql is similar, though Rust's language features enable a more composible 'code first' approach using macros.
+
+I believe the type of data we are consuming benefits from a strongly typed interface, and that compile-time client generation will significantly reduce the time-to-market of what will be already be a very complex front-end web app.
+
+## Prior Art
+
+From relatively superficial searching, I've not been able to find a comparable approach to using GraphQL for internal metrics / observability. However, there are [plenty of large companies](https://graphql.org/users/) using GraphQL for public API and very possibly for internal machinery that isn't public facing.
```

Comment:
Agreed, and especially in this case where the primary question is around the specific use case. I have no doubt that GraphQL is robust and well supported for its designed use case. The risk for us is whether that's also true for streaming data from an infrastructure tool instead of a more standard web app.

Reply:
I agree, but in the absence of companies using it in exactly the way we intend, I also think it's better than nothing.

I'm not sure what detail we're hoping to glean from prior art, or to a large extent, even the practical benefit. Do we care if another company is using JSON or Protobufs or GraphQL, if there are equally as many companies using the alternatives? This is also making a large assumption that other companies have chosen wisely/correctly, and that their circumstances apply to us.

I don't put a lot of stock into this research, unless it very obviously (and relatively recently) overlaps with our use-case.